File: lib/src/services/symptom_analyzer.py
import json
from huggingface_hub import InferenceClient
from typing import Dict, List, Optional

class SymptomAnalyzer:
    def __init__(self, api_token: str):
        """Initialize the SymptomAnalyzer with HuggingFace API token."""
        self.client = InferenceClient(token=api_token)
        self.model = "mistralai/Mistral-7B-Instruct-v0.1"  # You can change this to your preferred model
        
    async def analyze(self, symptoms_text: str) -> Dict:
        """
        Analyze the provided symptoms text and return structured analysis.
        
        Args:
            symptoms_text (str): Description of symptoms
            
        Returns:
            Dict containing potential conditions, recommendation, and urgency level
            
        Raises:
            ValueError: If the response cannot be parsed
            ConnectionError: If the API call fails
        """
        prompt = f"""You are an AI diagnostic assistant. Based on the following symptoms, 
        provide a structured analysis. Focus on being informative but cautious.
        
        Symptoms:
        {symptoms_text}
        
        Respond ONLY with a JSON object in this exact format:
        {{
            "potential_conditions": ["condition1", "condition2"],
            "recommendation": "brief medical advice",
            "urgency": "Low/Medium/High"
        }}
        
        Remember: This is for initial guidance only, not a final diagnosis."""
        
        try:
            response = await self.client.text_generation(
                prompt,
                model=self.model,
                max_new_tokens=500,
                temperature=0.3,
                return_full_text=False
            )
            
            return self._parse_response(response)
            
        except Exception as e:
            raise ConnectionError(f"Failed to get analysis: {str(e)}")
    
    def _parse_response(self, response: str) -> Dict:
        """
        Parse the LLM response into the required JSON structure.
        
        Args:
            response (str): Raw response from the LLM
            
        Returns:
            Dict with the parsed analysis
            
        Raises:
            ValueError: If response cannot be parsed into the required format
        """
        try:
            # Try to find JSON-like content in the response
            start_idx = response.find('{')
            end_idx = response.rfind('}') + 1
            if start_idx == -1 or end_idx == 0:
                raise ValueError("No JSON object found in response")
                
            json_str = response[start_idx:end_idx]
            parsed = json.loads(json_str)
            
            # Validate the required fields
            required_fields = ['potential_conditions', 'recommendation', 'urgency']
            if not all(field in parsed for field in required_fields):
                raise ValueError("Response missing required fields")
                
            # Validate urgency value
            if parsed['urgency'] not in ['Low', 'Medium', 'High']:
                raise ValueError("Invalid urgency value")
                
            # Ensure potential_conditions is a list
            if not isinstance(parsed['potential_conditions'], list):
                raise ValueError("potential_conditions must be a list")
                
            return parsed
            
        except json.JSONDecodeError as e:
            raise ValueError(f"Failed to parse response as JSON: {str(e)}")
            
        except Exception as e:
            raise ValueError(f"Error processing response: {str(e)}")

# Example usage:
# async def main():
#     analyzer = SymptomAnalyzer("your-hf-token-here")
#     result = await analyzer.analyze("Persistent headache for 3 days with sensitivity to light")
#     print(result)

File: lib/src/services/symptom_analyzer.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class SymptomAnalysis {
  final String potentialConditions;
  final String recommendation;
  final String urgency;

  SymptomAnalysis({
    required this.potentialConditions,
    required this.recommendation,
    required this.urgency,
  });

  factory SymptomAnalysis.fromJson(Map<String, dynamic> json) {
    return SymptomAnalysis(
      potentialConditions: json['potential_conditions'].join(', '),
      recommendation: json['recommendation'],
      urgency: json['urgency'],
    );
  }
}

class SymptomAnalyzer {
  final String apiToken;
  final String baseUrl;

  SymptomAnalyzer({
    required this.apiToken,
    this.baseUrl =
        'https://api-inference.huggingface.co/models/mistralai/Mistral-7B-Instruct-v0.1',
  });

  Future<SymptomAnalysis> analyzeSymptoms(String symptomsText) async {
    try {
      final response = await http.post(
        Uri.parse(baseUrl),
        headers: {
          'Authorization': 'Bearer $apiToken',
          'Content-Type': 'application/json',
        },
        body: jsonEncode({
          'inputs': _createPrompt(symptomsText),
          'parameters': {
            'max_new_tokens': 500,
            'temperature': 0.3,
            'return_full_text': false,
          },
        }),
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to analyze symptoms: ${response.statusCode}');
      }

      return _parseResponse(response.body);
    } catch (e) {
      throw Exception('Error analyzing symptoms: $e');
    }
  }

  String _createPrompt(String symptomsText) {
    return '''You are an AI diagnostic assistant. Based on the following symptoms, 
    provide a structured analysis. Focus on being informative but cautious.

    Symptoms:
    $symptomsText

    Respond ONLY with a JSON object in this exact format:
    {
        "potential_conditions": ["condition1", "condition2"],
        "recommendation": "brief medical advice",
        "urgency": "Low/Medium/High"
    }

    Remember: This is for initial guidance only, not a final diagnosis.''';
  }

  SymptomAnalysis _parseResponse(String responseBody) {
    try {
      // Extract JSON from response
      final jsonStart = responseBody.indexOf('{');
      final jsonEnd = responseBody.lastIndexOf('}') + 1;

      if (jsonStart == -1 || jsonEnd == 0) {
        throw FormatException('No JSON object found in response');
      }

      final jsonStr = responseBody.substring(jsonStart, jsonEnd);
      final parsed = jsonDecode(jsonStr);

      // Validate response structure
      if (!_isValidResponse(parsed)) {
        throw FormatException('Invalid response format');
      }

      return SymptomAnalysis.fromJson(parsed);
    } catch (e) {
      throw FormatException('Failed to parse response: $e');
    }
  }

  bool _isValidResponse(Map<String, dynamic> response) {
    return response.containsKey('potential_conditions') &&
        response.containsKey('recommendation') &&
        response.containsKey('urgency') &&
        response['potential_conditions'] is List &&
        response['urgency'] is String &&
        ['Low', 'Medium', 'High'].contains(response['urgency']);
  }
}
